### DISCUSSION TITLE
Trigger workflow for a pull request and for all subsequent commits in it

### DISCUSSION THREAD

**Discussion by raunaqgupta** [community/community#25161] [Actions]
<p>I want to run a workflow such that if a pull request has a certain filename pattern (amongst all commits when the pull request was opened or any subsequent commits before it’s merged ), it should run that workflow the moment a file was found matching the pattern.</p>
<p>I thought this would work but it doesn’t run for any commits inside the pull request that don’t have a matching file (which makes sense).</p>
<pre><code>on:
  push:
    paths:
      - '**.md'
</code></pre>
<p>In essence, the moment a pull request has a file matching the filetype (whether when it was opened or due to any subsequent commit), it should run that workflow for all commits pushed to that pull request subsequently till it’s merged.</p>
<p>Note: The only other potential alternative I can think of is to only run the workflow when the pull request is about to be merged and give an error if that’s possible.</p>
<p>Note: Another inquiry: is the event <strong>pull_request.edited</strong> fired when commits are pushed into it?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/raunaqgupta">@raunaqgupta</a> ,</p>
<p>I think you have some misunderstandings about <a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#push-event-push" rel="nofollow noopener"><strong>Push event</strong></a>, Push event will be triggered when you directly push commits to a repository branch or push tags to the repository.<br>
If you wants a workflow runs for pull request, in the YAML file of this workflow, you should set it runs on the events that are related to pull request (such as <a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request" rel="nofollow noopener"><strong>Pull request event</strong></a> and <a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-review-event-pull_request_review" rel="nofollow noopener"><strong>Pull request review event</strong></a>), not Push event.</p>
<p>For example:</p>
<pre><code>on:
  pull_request:
    types: [assigned, opened, synchronize, reopened]

on:
  pull_request_review:
    types: [edited, dismissed]
</code></pre>
<p>In an opening PR, if you want the subsequent commits pushed to this PR can trigger the workflow, you can set the workflow runs on the activity type  <strong>pull_request.synchronize</strong>.</p>
<pre><code>on:
  pull_request:
    types: [opened, synchronize]
</code></pre>
<p>By default, a workflow only runs when a <strong>pull_request</strong>’s activity type is <strong>opened</strong> , <strong>synchronize</strong> , or <strong>reopened</strong>. So you can just set it like this:</p>
<pre><code>on: pull_request
</code></pre>
<p>It equals the below configuration:</p>
<pre><code>on:
  pull_request:
    types: [opened, synchronize, reopened]
</code></pre>
<p>About the path filters on Pull request event, you need to know:</p>
<p>Every time you push a new commit to the opening PR, the Actions app will read all the modified files in this PR from the first to last. Once at least one modified file matches the path filters to trigger the workflow, the workflow will be triggered.</p>
<p>That means, if in the previous commits have modified files match the path filters, although in a subsequent commit you only modify the files that do not match the path filters, the workflow still will be triggered by the activity type  <strong>pull_request.synchronize</strong>.</p>
<p>In addition, the activity type  <strong>pull_request.edited</strong>  will trigger the workflow when the users edit somethings on the PR, such as changes the PR’s title, changes the PR’s description, or some other information displayed on the web page of the PR.</p>

**Comment by jenschelkopf** [comment]
<p>Thanks <a class="mention" href="/brightran">@brightran</a>!</p>
<p><a class="mention" href="/raunaqgupta">@raunaqgupta</a> to summarize, you can accomplish your goal by using the <code>opened</code> and <code>synchronize</code> activity types for the <code>pull_request</code> event. The <code>opened</code> type is obvious :grinning: The <code>synchronize</code> type happens whenever a commit is pushed to the PR. I’d also recommend using <code>reopened</code> in case a PR is closed and then later opened again.</p>
<p>Here’s an example:</p>
<pre><code>on:
  pull_request:
    types: [opened, synchronize, reopened]
    paths: '**.md'
</code></pre>
<p>This workflow will not run if a PR is opened, or pushed to, without any markdown changes. One a markdown change is pushed to the PR, it will run with every subsequent push (even when a particular commit does not contain any markdown files).</p>

**Reply by ganeshgk** [reply]
opened, synchronize & reopened are by default used even if types is not specified, so the below also is fine 

```
on:
  pull_request:
    paths: '**.md'
```

**Comment by raunaqgupta** [comment]
<p>Thanks for the explanation. I was aware that push would not give me what I needed and Github’s event documentation is pretty sparse in explaining what those sub-events actually mean or when are they fired. I’m talking about this particular page: <a href="https://developer.github.com/v3/activity/events/types/#pullrequestevent" rel="nofollow noopener">https://developer.github.com/v3/activity/events/types/#pullrequestevent</a> Is there any other page which I have clearly been unable to find where the sub-events (especially <strong>synchronize</strong> ) are clearly explained?</p>
<p>Anyway, <strong>synchronize</strong> is exactly what I need. Cheers.</p>
